ciongke/cmd/test-pr/main.go

```diff
@@ -0,0 +1,203 @@
+	author     = flag.String("author", "", "Author of the PR.")
+	bucket     = flag.String("bucket", "", "Bucket to write results to.")
+	timeoutMin = flag.Int("timeout", 10, "Timeout for the build step in minutes.")
+)
+
+const (
+	jobName   = "test-infra"
+	workspace = "/workspace"
+	repo      = "https://github.com/kubernetes/test-infra"
+	// TODO: Use a cluster-local git mirror so that we don't hit GitHub so often.
+)
+
+func main() {
+	flag.Parse()
+
+	// Fail early if these aren't set.
+	if *pr == 0 || *author == "" || *bucket == "" {
+		flag.Usage()
+		log.Fatalf("Must provide pr, author, and bucket.")
+	}
+
+	// Set logging to write to both stdout and the build log file.
+	buildLogPath := filepath.Join(workspace, "build-log.txt")
+	file, err := os.Create(buildLogPath)
+	if err != nil {
+		// Quit here, we won't be able to upload the build log anyway.
+		log.Fatalf("Could not open log file (%s): %s", buildLogPath, err)
+	}
+	log.SetOutput(io.MultiWriter(file, os.Stdout))
+
+	startTime := time.Now()
+	result := gcs.Success
+
+	clonePath := filepath.Join(workspace, "repo")
+	commit, trusted, mergeable, err := checkoutPR(repo, clonePath, *author, *pr)
+	if trusted != nil {
+		// TODO: Status update to "Add author to .test-whitelist.txt"
+		log.Printf("Cannot proceed: %s", trusted)
+		result = gcs.Failure
+	} else if mergeable != nil {
+		// TODO: Status update to "Needs rebase"
+		log.Printf("Cannot proceed: %s", mergeable)
+		result = gcs.Failure
+	} else if err != nil {
+		log.Printf("Error preparing git: %s", err)
+		result = gcs.Failure
+	}
+
+	buildNumber, err := gcs.IncrementAndGetBuildNumber(*bucket, jobName)
+	if err != nil {
+		log.Printf("Could not get build number: %s", err)
+		result = gcs.Failure
+	}
+	buildPath := jobName + "/" + strconv.Itoa(buildNumber)
+	if err = gcs.UploadStarted(*bucket, buildPath, commit); err != nil {
+		log.Printf("Could not upload started.json: %s", err)
+		result = gcs.Failure
+	}
+
+	// Only run tests if previous steps were successful.
+	if result == gcs.Success {
+		// TODO: Status update to "Test started"
+		timeout := time.Duration(*timeoutMin) * time.Minute
+		if err = build(clonePath, timeout); err == nil {
+			// TODO: Status update to "Success"
+			log.Printf("Build succeeded after %v", time.Since(startTime))
+		} else {
+			// TODO: Status update to "Failure"
+			log.Printf("Build failed after %v: %s", time.Since(startTime), err)
+			result = gcs.Failure
+		}
+	}
+
+	if err = gcs.UploadFinished(*bucket, buildPath, result); err != nil {
+		log.Printf("Could not upload finished.json: %s", err)
+	}
+	if err = gcs.UploadBuildLog(*bucket, buildPath, buildLogPath); err != nil {
+		log.Printf("Could not upload build log: %s", err)
+	}
+}
+
+// checkoutPR returns the commit hash, an error if the author is untrusted, an
+// error if the PR can't be merged, and an error if anything else fails.
+func checkoutPR(repoPath, clonePath, prAuthor string, prNumber int) (string, error, error, error) {
```

Comment:
I made it just return `(bool, error)`: if it can be merged and if there was some other error.